postgresql-common-1/pg_default.1
.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.TH PG_DEFAULT 1 "October 2003" "Debian Project" "Debian Linux"
.SH NAME
pg_default \- Sets cluster and database for PostgreSQL commands
.SH SYNOPSIS
.P
.B pg_default
.BR -a " | "
.BI -u " user "
.RB "| " -g
.I " group"
.RB [ -f ]
.BI -c " cluster"
[
.BI -d " database"
]
.P
.B pg_default
[
.BI -c " cluster"
[
.BI -d " database"
]] | [-V]
.SH OPTIONS
.TP
.BI -u " user"
Specify the Unix login for which the defaults are to be set.  
This option may be used only by root or the administrator.
.TP
.BI -g " group"
Specify the Unix group for which the defaults are to be set. 
This option may be used only by root or the administrator.
.TP
.B -a
Specify the default choice for all users, that will apply if no more
specific match of user or group is found.
This option may be used only by root or the administrator.
.TP
.B -f
Set the FORCED flag in
.IR user_clusters ,
so that the user is unable to override the choice made for him.
This option may be used only by root or the administrator.
.TP
.BI -c " cluster"
Specify the name of the cluster (in 
.IR cluster_ports )
to which the user should connect by default when a PostgreSQL command is run.
.TP
.BI -d " database"
Specify the name of the database within the specified cluster to which the
user should connect by default.  If
.I database
is *, the connection will be to a database whose name is the same as the
user's login.
.TP
.B -V
When
.B pg_default
is used to report the current settings, this option makes
it report the default system settings as well as the user's own settings,
or it reports that the system settings cannot be overridden.
.SH DESCRIPTION
.P
When 
.B pg_default
is run with no options, it reports the default database
and cluster for the current user.
If
.B -c
is specified,
.B pg_default
sets the PostgreSQL cluster for a user, and the database too, if
.B -d
is also given.
.P
The superuser (root) or the PostgreSQL administrator (postgres) may
use the 
.B -u
and
.B -g
options to set the defaults for another user and the 
.B -a
option to set the default cluster and database.
The 
.B -f
option is used to set the FORCED flag in
.IR user_clusters ,
so that the user
is not permitted to override the administrator's choices for him.
.P
Other users can set only their own options, which are stored in
.IR ~/.postgresqlrc .
.SH FILES
.TP
.I /etc/postgresql/user_clusters 
stores the default cluster and database for users and groups as set by
the administrators.  By setting the FORCED flag for a user or group, it
is possible for an administrator to override a
user's choices so that the user can connect only to the cluster and
database selected for him.
.TP
.I $HOME/.postgresqlrc
stores defaults set by the user himself.
.SH BUGS
The FORCED flag is meant to protect naive users from stupid mistakes.  It
is not a security measure, since it can easily be bypassed by entering the
desired pathname and setting PGPORT directly.
.SH SEE ALSO
.BR pg_exec (1),
.BR pg_wrapper (1)
.SH AUTHOR
Oliver Elphick <olly@lfix.co.uk>


debian/NEWS
postgresql-common (158) UNRELEASED; urgency=medium

  This postgresql-common release introduces several changes in the defaults
  used for newly created database clusters, all configurable in
  /etc/postgresql-common/createcluster.conf:

  * The default log_line_prefix is now '%t [%p-%l] %q%u@%d ' instead of '%t '.
  * stats_temp_directory aka pg_stat_tmp is set to a subdirectory of
    /var/run/postgresql by default.
  * The use of SSL by default is (pre-)configurable via debconf and
    createcluster.conf.

  If the logging_collector or syslog is being used, pg_lsclusters will show
  the filename (or "syslog") instead of just "custom" in the Log file column.

 -- Christoph Berg <christoph.berg@credativ.de>  Thu, 26 Jun 2014 13:23:52 +0200


postgresql-common-1/pg_default.c
#/*
   pg_default.c
  
   Set defaults for pg_wrapper.
  
   Copyright (c) 2003 Oliver Elphick <olly@lfix.co.uk>
   Licence:  GNU Public Licence v.2 or later

 */

#include "pg_wrapper.h"
#include "pg_vars.h"

#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <pwd.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>

void show_default(void);
void set_own_default(void);
void set_all_default(void);
void set_group_default(void);
void set_user_default(void);
void set_defaults(void);

/********************************************************************
 Set the defaults requested 
 ********************************************************************/
void set_defaults (void) {
  int uid;

  if (username[0] != '\0' || groupname[0] != '\0' || set_default) {
    /* These options require root privilege or postgres */
    if ((uid = geteuid())) /* id 0 is root */ {
      /* Not root; is it postgres? */
      if (getpwnam("postgres")->pw_uid != uid) {
	fprintf(stderr, "Only the superuser or the user 'postgres' can do that\n");
        exit(LOC_ERR_PERM);
      }
    }
    if (clustername[0] != '\0') {
      if (!valid_cluster(clustername)){
	fprintf(stderr,"Unknown cluster %s\n", clustername);
	exit(LOC_ERR_SYNTAX);
      }
    } else {
      fprintf(stderr,"Cluster not specified\n");
      exit(LOC_ERR_SYNTAX);
    }
    if (username[0] != '\0')
	set_user_default();
    if (groupname[0] != '\0')
	set_group_default();
    if (set_default)
	set_all_default();
  } else {
    if (force || release) {
      fprintf(stderr, "Only the superuser can use -f or -r\n");
      exit(LOC_ERR_PERM);
    }
   if (clustername[0] == '\0' && dbname[0] == '\0') {
     show_default();
   } else {
     set_own_default();
   }
  }
}

void set_user_default(void) {
  write_cluster_line(username, "*", clustername, dbname, force);
}

void set_group_default(void) {
  write_cluster_line("*", groupname, clustername, dbname, force);
}

void set_all_default(void) {
  write_cluster_line("*", "*", clustername, dbname, force);
}

void set_own_default(void) {
  write_rc( clustername, dbname);
}

void show_default(void) {
  char buf[PATHLEN];  /* temporary buffer */
  char ugname[BUFSIZ] = "";
  char sdcluster[BUFSIZ] = "";
  char sddb[BUFSIZ] = "";
  char sdf[BUFSIZ] = "";
  char owncluster[BUFSIZ] = "";
  char owndb[BUFSIZ] = "";
  int ugno, frc;
  struct passwd *pwd;
  
  find_settings(buf);
  if (!(strncmp(dbname, "*", 2)) || dbname[0] == '\0') {
    pwd = getpwuid(getuid());
    strncpy(dbname, pwd->pw_name, BUFSIZ);
  }
  printf("Cluster: %s; database: %s\n", clustername, dbname);
 
  if (verbose) {
    get_sys_default_info(&ugno, ugname, sdcluster, sddb, &frc);
    if (frc) {
      strcpy(sdf, "FORCED");
    }

    switch (ugno) {
    case 1:
      printf("\nSystem defaults for user %s --\n"
	     "  cluster: %s; database: %s   %s\n", 
	     ugname, sdcluster, sddb, sdf);
      break;
    case 2:
      printf("\nSystem defaults for group %s --\n"
	     "  cluster: %s; database: %s   %s\n", 
	     ugname, sdcluster, sddb, sdf);
      break;
    case 4:
      strcpy(sdf, "(cluster for port 5432)");
    case 3:
      printf("\nSystem defaults for unspecified users --\n"
	     "  cluster: %s; database: %s   %s\n", 
	     sdcluster, sddb, sdf);
      break;
    case 0:
      printf("\nNo system default and no cluster for port 5432\n");
    }
    if (!frc) {
      get_user_default_info(owncluster, owndb);
      printf("%s/.postgresqlrc:\n"
	     "  cluster: %s; database: %s\n",
	     getenv("HOME"), owncluster, owndb);

      owncluster[0] = '\0';
      if ((getenv("PGCLUSTER")) != NULL) {
	strncpy(owncluster, getenv("PGCLUSTER"), BUFSIZ);
      }
      owndb[0] = '\0';
      if ((getenv("PGDATABASE")) != NULL) {
	strncpy(owndb, getenv("PGDATABASE"), BUFSIZ);
      }
      if (owncluster[0] != '\0' || owndb[0] != '\0') {
	printf("Environment:\n  PGCLUSTER: %s; PGDATABASE: %s\n",
	       owncluster, owndb);
      }
    }
  }
}
